File: summary/rust/15_k_means.rs
use std::cmp::Ordering;

#[derive(Debug, Copy, Clone)]
struct Point {
    x: f64,
    y: f64,
}

impl Point {
    fn new(x: f64, y: f64) -> Point {
        Point { x, y }
    }

    fn distance(&self, other: &Point) -> f64 {
        ((self.x - other.x).powi(2) + (self.y - other.y).powi(2)).sqrt()
    }
}

#[derive(Debug)]
struct Cluster {
    centroid: Point,
    points: Vec<Point>,
}

impl Cluster {
    fn new(centroid: Point) -> Cluster {
        Cluster {
            centroid,
            points: Vec::new(),
        }
    }

    fn update_centroid(&mut self) {
        let sum_x: f64 = self.points.iter().map(|p| p.x).sum();
        let sum_y: f64 = self.points.iter().map(|p| p.y).sum();
        let num_points = self.points.len() as f64;
        self.centroid.x = sum_x / num_points;
        self.centroid.y = sum_y / num_points;
    }

    fn clear_points(&mut self) {
        self.points.clear();
    }
}

fn k_means(points: &[Point], k: usize, max_iterations: usize) -> Vec<Cluster> {
    let mut clusters: Vec<Cluster> = Vec::new();

    // Initialize clusters with random centroids
    for i in 0..k {
        let centroid = points[i];
        clusters.push(Cluster::new(centroid));
    }

    for _ in 0..max_iterations {
        // Assign points to clusters
        for point in points {
            let mut min_distance = f64::MAX;
            let mut closest_cluster_index = 0;
            for (i, cluster) in clusters.iter().enumerate() {
                let distance = point.distance(&cluster.centroid);
                if distance < min_distance {
                    min_distance = distance;
                    closest_cluster_index = i;
                }
            }
            clusters[closest_cluster_index].points.push(*point);
        }

        // Update centroids
        for cluster in &mut clusters {
            cluster.update_centroid();
        }

        // Clear points from clusters
        for cluster in &mut clusters {
            cluster.clear_points();
        }
    }

    clusters
}

fn main() {
    // Sample data points
    let points = vec![
        Point::new(1.0, 1.0),
        Point::new(2.0, 1.0),
        Point::new(1.5, 2.0),
        Point::new(8.0, 8.0),
        Point::new(7.0, 9.0),
        Point::new(9.0, 8.0),
    ];

    // Number of clusters
    let k = 2;

    // Maximum iterations
    let max_iterations = 100;

    // Run k-means algorithm
    let clusters = k_means(&points, k, max_iterations);

    // Output cluster centroids and points
    for (i, cluster) in clusters.iter().enumerate() {
        println!("Cluster {}: {:?}", i + 1, cluster);
    }
}

File: summary/c++/15_k_means.cpp
#include <vector>
#include <cmath>
#include <limits>
#include <algorithm>
#include <iostream>

double euclidean_distance(const std::vector<double>& point1, const std::vector<double>& point2) {
    double sum = 0.0;
    for (size_t i = 0; i < point1.size(); ++i) {
        sum += (point1[i] - point2[i]) * (point1[i] - point2[i]);
    }
    return std::sqrt(sum);
}

void k_means(const std::vector<std::vector<double>>& data, int k, std::vector<std::vector<double>>& centroids, std::vector<std::vector<std::vector<double>>>& clusters) {
    // Initialize centroids randomly
    centroids.resize(k);
    std::vector<int> used_indices;
    for (int i = 0; i < k; ++i) {
        int index;
        do {
            index = rand() % data.size();
        } while (std::find(used_indices.begin(), used_indices.end(), index) != used_indices.end());
        used_indices.push_back(index);
        centroids[i] = data[index];
    }

    bool changed;
    do {
        // Create clusters
        std::vector<std::vector<std::vector<double>>> new_clusters(k);
        for (const auto& point : data) {
            int closest_centroid_index = 0;
            double min_distance = std::numeric_limits<double>::max();
            for (int i = 0; i < k; ++i) {
                double distance = euclidean_distance(point, centroids[i]);
                if (distance < min_distance) {
                    min_distance = distance;
                    closest_centroid_index = i;
                }
            }
            new_clusters[closest_centroid_index].push_back(point);
        }

        // Update centroids
        changed = false;
        for (int i = 0; i < k; ++i) {
            std::vector<double> new_centroid(data[0].size(), 0.0);
            for (const auto& point : new_clusters[i]) {
                for (size_t j = 0; j < point.size(); ++j) {
                    new_centroid[j] += point[j];
                }
            }
            for (size_t j = 0; j < new_centroid.size(); ++j) {
                new_centroid[j] /= new_clusters[i].size();
            }
            if (euclidean_distance(new_centroid, centroids[i]) > 0.001) {
                centroids[i] = new_centroid;
                changed = true;
            }
        }
        clusters = new_clusters;
    } while (changed);
}

int main() {
    std::vector<std::vector<double>> data = {{1.0, 2.0}, {1.5, 1.8}, {5.0, 8.0}, {8.0, 8.0}, {1.0, 0.6}, {9.0, 11.0}};
    int k = 2;
    std::vector<std::vector<double>> centroids;
    std::vector<std::vector<std::vector<double>>> clusters;
    k_means(data, k, centroids, clusters);

    for (int i = 0; i < k; ++i) {
        std::cout << "Cluster " << i + 1 << ": Centroid -> (";
        for (size_t j = 0; j < centroids[i].size(); ++j) {
            std::cout << centroids[i][j] << (j < centroids[i].size() - 1 ? ", " : "");
        }
        std::cout << ")" << std::endl;
        std::cout << "Points:" << std::endl;
        for (const auto& point : clusters[i]) {
            std::cout << "(";
            for (size_t j = 0; j < point.size(); ++j) {
                std::cout << point[j] << (j < point.size() - 1 ? ", " : "");
            }
            std::cout << ")" << std::endl;
        }
    }
    return 0;
}
